src/librbd/mirror/snapshot/CreatePrimaryRequest2.h

```diff
@@ -0,0 +1,152 @@
+#define CEPH_LIBRBD_MIRROR_SNAPSHOT_CREATE_PRIMARY_REQUEST2_H
+
+#include "include/buffer.h"
+#include "include/rados/librados.hpp"
+#include "cls/rbd/cls_rbd_types.h"
+#include "librbd/internal.h"
+#include "librbd/mirror/snapshot/Types.h"
+
+#include <string>
+#include <set>
+
+struct Context;
+
+namespace librbd {
+
+struct ImageCtx;
+
+namespace mirror {
+namespace snapshot {
+
+template <typename ImageCtxT = librbd::ImageCtx>
+class CreatePrimaryRequest2 {
+public:
+  static CreatePrimaryRequest2 *create(std::vector<ImageCtxT *>&image_ctxs,
+                                       std::vector<std::string> &global_image_ids,
+                                       std::vector<uint64_t> &clean_since_snap_ids,
+                                       uint64_t snap_create_flags,
+                                       uint32_t flags,
+                                       const std::string &group_snap_id,
+                                       std::vector<uint64_t> &snap_ids,
+                                       Context *on_finish) {
+    return new CreatePrimaryRequest2(image_ctxs, global_image_ids,
+                                     clean_since_snap_ids, snap_create_flags,
+                                     flags, group_snap_id, snap_ids, on_finish);
+  }
+
+  static CreatePrimaryRequest2 *create(std::vector<ImageCtxT *>&image_ctxs,
+                                       std::vector<std::string> &global_image_ids,
+                                       std::vector<uint64_t> &clean_since_snap_ids,
+                                       uint64_t snap_create_flags,
+                                       uint32_t flags,
+                                       std::vector<uint64_t> &snap_ids,
+                                      Context *on_finish) {
+    return new CreatePrimaryRequest2(image_ctxs, global_image_ids,
+                                    clean_since_snap_ids, snap_create_flags,
+                                    flags, {}, snap_ids, on_finish);
+  }
+
+/*
+  CreatePrimaryRequest2(ImageCtxT *image_ctx,
+                       const std::string& global_image_id,
+                       uint64_t clean_since_snap_id,
+                       uint64_t snap_create_flags, uint32_t flags,
+                       const std::string &group_snap_id, uint64_t *snap_id,
+                       Context *on_finish);
+*/
+
+  CreatePrimaryRequest2(std::vector<ImageCtxT *> &image_ctxs,
+                        std::vector<std::string> &global_image_ids,
+                        std::vector<uint64_t> &clean_since_snap_ids,
+                        uint64_t snap_create_flags,
+                        uint32_t flags,
+                        const std::string &group_snap_id,
+                        std::vector<uint64_t> &snap_ids,
+                        Context *on_finish);
+
+  void send();
+
+private:
+  /**
+   * @verbatim
+   *
+   * <start>
+   *    |
+   *    v
+   * GET_MIRROR_PEERS
```

Comment:
The state diagram doesn't include `notify_quiesce` , `notify_unquiesce`, `acquire_exclusive_lock()` and `release_exclusive_lock()`? Can that be added?

Reply:
Done.